### ISSUE TITLE
Add support for refresh tokens

### ISSUE THREAD

**Issue by marshallswain** [feathersjs/feathers#1337]
We currently allow getting a new token by posting a valid auth token to `<loginEndpoint>/refresh`.  Refresh tokens have a slightly different workflow as explained here:
https://auth0.com/learn/refresh-tokens

**Comment by philoez98** [comment]
Is there any mindful way of using refresh tokens in feathers right now? Any plans of adding support for them?

**Comment by bujji1** [comment]
Hello All ,

Looks like it is still not available ( or Am I missing some thing ) . Could you please let us know when it will be available ?  I see there is passport refresh token repository available . Any one tried this ?
https://github.com/fiznool/passport-oauth2-refresh

**Comment by khanshakeeb** [comment]
Hello all,
Can you anyone help me to understand how token can be refresh for the google strategy? Because I will not have password in google login scenario?

Thanks

**Comment by th0r** [comment]
> Another approach is to store refresh token inside accessToken's payload

So anyone who have **even one** of your **even expired** access tokens will be able to easily generate an endless count of a new access tokens or am I miss something?

Refresh tokens should be **securely** stored on the client because of exact same reason! 

> @deiucanta the good news is that we kept this feature in mind while we designed auth@1.0. I don't think it will be long before we get it in place and documented.

This was posted in 2016. Guys, do you still have plans to support this must-have feature?

**Comment by daffl** [comment]
No. You can't do anything with an expired token. Also, refresh tokens are much more easily possible in [the v4 prerelease](https://blog.feathersjs.com/feathers-spring-2019-news-and-v4-beta-test-4c023c283e87). A cookbook entry for how to do it will be part of the final release.

**Comment by th0r** [comment]
> Also, refresh tokens are much more easily possible in [the v4 prerelease](https://blog.feathersjs.com/feathers-spring-2019-news-and-v4-beta-test-4c023c283e87).

Is there an approximate release date? 

> A cookbook entry for how to do it will be part of the final release.

Until this guide is released, could you explain in a few words how it can be done in v4 prerelease?

**Comment by th0r** [comment]
@daffl ping

**Comment by ellisadigvom** [comment]
Is there an official way to do this yet? v4 is here and I don't see anything in the docs

**Comment by MichaelErmer** [comment]
@daffl could you elaborate how this is achievable with 4.0 and without hacks to the authentication service?

**Comment by sarkistlt** [comment]
@MichaelErmer as a workaround you can use local or any custom strategy to renew jwt, not ideal, but works fine for internal communication, let's say between worker and api.

```js
function reAuth() {
  return async (ctx) => {
    if (
      ctx.error
      && ctx.error.code === 401
      && !ctx.params.skipAuthHook
    ) {
      try {
        ctx.params.skipAuthHook = true;
        await ctx.app.authenticate(apiLocalCreds);
        delete ctx.error;
      } catch (err) {
        ctx.error = err;
      }
    }
    return ctx;
  };
}

client
  .configure(rest(apiHost).superagent(superagent))
  .configure(auth(authConfig))
  .hooks({ error: [reAuth()] });
```

**Comment by m0dch3n** [comment]
Currently I'm using this `after` hook in v4 `authentication`, to update my accessToken after 20 days...

````javascript
const {DateTime} = require('luxon')
const renewAfter = {days: 20}

module.exports = () => {
  return async context => {
    if (
      context.method === 'create' &&
      context.type === 'after' &&
      context.path === 'authentication' &&
      context.data && context.data.strategy === 'jwt' &&
      context.result &&
      context.result.accessToken) {
      // check if token needs to be renewed
      const payload = await context.app.service('authentication').verifyAccessToken(context.result.accessToken)
      const issuedAt = DateTime.fromMillis(payload.iat * 1000)
      const renewAfter = issuedAt.plus(renewAfter)
      const now = DateTime.local()
      if (now > renewAfter) {
        context.result.accessToken = await context.app.service('authentication').createAccessToken({sub: payload.sub})
      }
    }
    return context
  }
}
````

It's important to have this hook in `after` and as last hook, so that all the verifications etc have passed

**Comment by PowerMogli** [comment]
Any plans to integrate refresh tokens ins feathers?

**Comment by 1valdis** [comment]
I second that question one message earlier.

**Comment by rdewolff** [comment]
Am wondering about the refresh token workflow as well. Is the solution drafted by @m0dch3n a good practice? Should we implement it another way ?

**Comment by m0dch3n** [comment]
The whole refreshToken workflow in my opion only protects only a little bit against man in the middle attacks, so that if the middle man steals the accessToken he can at least not refresh it and have infinite access to the ressources. 

It does not protect against XSS, because in that case, the attacker is able to steal anything stored on the client side. So also the refreshToken...

The problem now is, that if you make your accessToken expiration time too small (i.e. 5 minutes), you also have too refresh it more often. The man in the middle only needs to listen during 5 minutes to the clients requests in order to intercept the refreshToken then...  If you make the expiration longer, he has longer access with just the accessToken...

Honestly if some client tells me, his access got stolen, I need to blacklist accessToken **AND** refreshToken anyway to be sure. So I'm forced to make a DB request on each request anyway. 

In my case, when I'm aware of such a case, I blacklist all the accessTokens from the last 40 days, because my accessTokens have a validity of 40 days...

**Comment by rdewolff** [comment]
Using HTTPS request makes man in the middle attacks really difficult. Aren't you using HTTPS requests?

**Comment by m0dch3n** [comment]
Of course I'm using https, but there are 3 possibilities to steal the accessToken. First is on client side (XSS i.e.), second on transport (man in the middle), and third on server side. 

On client and on transport, I'm only half responsible for the security, and the other half is the client, which is not totally under my control. But I can help the client, to avoid security risks, by making XSS impossible and by securing the transport with https...

The goal of a refreshToken is, to make the expiration of an accessToken shorter **AND to not transmit a longer or infinit valid token on EACH request** 

So the only security it brings, is that from 100 requests i.e, you don't make all the 100 vulnerable on transport, but only 1 request

So basically a man in the middle attack, can't be protected by a refeshToken and of course not by an XSS...

**Comment by jackywxd** [comment]
I think refresh token is a must support feature and it is not automatically renew existing access token. Access token is stateless and won’t be stored in server side. The down side is it is valid forever! The longer of access token, more risk is imposed. But if access token is too short, then your users have to login quite often, that will greatly impact usability.
That’s where refresh token comes in, the token used to refresh access token and it is a long live token. When access token expired, client can use refresh token to get a new access token, and that’s the only purpose of refresh token.
Refresh token is revokable in case of user account has been compromised. And that’s the big difference between access token and refresh token. To revoke issued refresh token, server must store all issued refresh tokens. In other words, refresh token is stateful. Server needs to know which one is valid which one is invalid.
To properly implement refresh token, we need some sort of token store to persist refresh token. We also need to implement at least three flows:
Refresh token validation
Refresh access token with valid refresh token
Revoke compromised user’s refresh token
There are other management functionalities also nice to have such as token usage stats.
Above is my current understanding regarding how to implement refresh token. It is not easy but it definitely necessary to build a more secure system.

**Comment by jackywxd** [comment]
It turns out Feathers already built-in all functionalities/modules required to properly implement refresh-tokens:

1. Refresh-token store: can be easily supported by Feathers Service.
2. Issuing and validating refresh token: can just re-used existing JWT support which built-in AuthenticationService.

Based on the work done by TheSinding (https://github.com/TheSinding/authentication-refresh-token), I implemented my own version of refresh-tokens with one custom service and three hooks (https://github.com/jackywxd/feathers-refresh-token) which enables basic refresh-tokens functionalities:

1. Issue refresh-token after user authentication successfully;
2. Refresh access token with a valid JWT refresh-token;
3. Logout user by deleting the refresh-token

While fully leverage existing code base in Feathres, the actually coding effort is minimum, and it integrates with current Feathers architecture nicely. It proves that current Feathers architecture is very extendable.

But a full feature of Refresh-token also requires support at Client side, such as store refresh-token in client side, reAuthenticate user after access-token expiration, logout user with refresh-token.

After review the source code of feathers-authentication and authentication-client, I believe refresh-token could be tapped into existing Features code based to allow  turning on refresh-token support as easy as turning on authentication.

I already ported my hooks version refresh-token code base into @feathersjs/authentication. Next I would try to make change on authentication-client to enable client side features. My ultimate goal is to enable refresh-token support in both server and client side.

**Comment by bwgjoseph** [comment]
My question/concern is how would the refresh token be stored in the client? 

See https://auth0.com/blog/securing-single-page-applications-with-refresh-token-rotation/

> Unfortunately, long-lived RTs are not suitable for SPAs because there is no persistent storage mechanism in a browser that can assure access by the intended application only. As there are vulnerabilities that can be exploited to obtain these high-value artifacts and grant malicious actors access to protected resources, using refresh tokens in SPAs has been strongly discouraged.

See https://afteracademy.com/blog/implement-json-web-token-jwt-authentication-using-access-token-and-refresh-token

> So, what is the best possible place to store the tokens securely? You can read more about it on the internet if you are passionate to achieve completely secure storage. Some of the solutions are ideal but not very practical. Practically I would store it in the Cookies with httpOnly and Secure flags. It is not 100 percent secure but it gets the job done.

See this long discussion on cookie - https://github.com/feathersjs-ecosystem/authentication/issues/132 too

**Comment by sarkistlt** [comment]
@bwgjoseph I would suggest to use regular express session, and store all tokens there, instead of client side. That's what I do and works perfectly fine with all types of apps including SPA

**Comment by bwgjoseph** [comment]
@sarkistlt You mean to say to store all client JWT token on server-side? Any reference/article material for that? I'm not exactly sure how the process would be like. So what does client sent, when they request for data (CRUD)?

**Comment by sarkistlt** [comment]
@bwgjoseph same as always, cookie, just add middlewere before registerring your services:

```js
app.use('* | [or specific rout]', session(sess), (req, res, next) => {
      req.feathers.session = req.session || {};
      next();
    });
```

then in your server, for let's customer login service, when customer is authenticated, you just store token in the session like `ctx.params.session.token = token`, where token is your JWT access or refresh token, depends on your application logic.
And with any new request from client you will check if token exist in the session and will use it for authentication. This is much safer and secure approach, since none of the tokens are exposed on the client side at all.

**Comment by daffl** [comment]
This has been discussed _a lot_ before (I also added an [entry to the FAQ](https://docs.feathersjs.com/help/faq.html#why-are-you-using-jwt-for-sessions)) and it is not necessarily more secure to store a token in a session. If someone gets access to your page to be able to execute scripts they have also hijacked the session and can make authenticated requests anyway.

Hence it is usually ok to store a token in e.g. localStorage (which only the current page has access to as well) and it also works seamlessly with other non-browser platforms (like native mobile apps, server-to-server etc.) __and websockets__ (I can't stress enough how painful it is to make websockets work seamlessly and securely with HTTP cookies - my life has been a lot easier since we stopped trying to do so). In general, a refresh token should be revokable though since it is usually a lot more long lived.

Either way, a pull request for this would be very welcome, I find it makes ironing out the details a lot easier.

**Comment by TheSinding** [comment]
@jackywxd - Great job, took a brief look at it and it seems like some great additions. 
Is there anyway of making it easier for the developer implementing this? 
Like integrating the hooks you've created into the library, so we wouldn't need to add the hooks later ? 

I think you should create a pull request and we could have the discussion there.

**Comment by sarkistlt** [comment]
@daffl yes agree, especially with WS. And if both client and backend build by you or your team, yes it's best to just use JWT and avoid extra dependancies and complexity in your application.
But in some cases, for example when building storefront REST-API that will be used by 3rd party companies / developers, it's easier to use regular session, and ask developers to include credentials with their requests then to describe how to retrieve access (and refresh) tokens, store it, and pass it with each requests. Which handled perfectly by feathers client, but in most cases when developer is not familiar with how backend built, they will use `request, superagent, fetch or axios` to connect their application to the backend. At least in my case this was the main reason to move storefront part of API to work with regular sessions instead of JWT directly.

**Comment by TheSinding** [comment]
But wouldn't that be the design decision and responsibility, of the maker of said storefront, to implement into their own API and the properly document this feature, instead of "forcing" this decision onto the community ?

**Comment by sarkistlt** [comment]
@TheSinding I think we can say 'forcing' about less commonly used approach, not about cookie which have been (and still is) most commonly used approach to manage user sessions. 

And yes it is a design decision made after developers feedback that have been using the API. When you are running multi-tenant system or API that used by multiple teams, sometime best is to follow general industry practice to avoid additional confusion and extra time spent by developers, especially if alternative solution doesn't provide any advantage for end user.

Again to be clear, using JWT is great, and way easier to utilize especially for real time API and that's what we are using in 90% of cases + you don't need to run redis or something else to manage cookie sessions. 
But there some exceptional cases where it may not be the best chose, I brought example of that situation in my previous comment.

**Comment by jackywxd** [comment]
Thanks for all your feedback! JWT vs session is a different topic we can discuss separately. 

I think one thing we all agree is that access token + refresh token is a much more secure solution than merely access token. It has been widely adopted by major Internet giants. It is fair to say the Feathers community would love to see Feathers main code base to provide built-in support for refresh token. Actually it surprised me when I first realized that Feathers doesn't support refresh token. 

I have been using AWS cognito in couple of my projects, AWS Amplify will save three tokens in localStorage: ID token, access token and refresh token, Amplify handles all the dirty-works related to token management,  and provides couple APIs that enables easy and straight forward interface working with Cognito backend. I would like to see similar development experience with Feathers.

**Comment by jackywxd** [comment]
@TheSinding Thanks for your previous great work! 

Because existing authentication is implemented as normal service, we could easily enable refresh-token support by extending the class and adding couple hooks:

`  this.hooks({
      after: {
        create: [issueRefreshToken(), connection('login'), event('login')],
        remove: [logoutUser(), connection('logout'), event('logout')],
        patch: [refreshAccessToken()],
      },`

Just like we turn on Authentication by using the CLI, we could offer the similar option in CLI for refresh-token support. developer can simply answer YES, the CLI then automatically creating a customer "refresh-tokens" service and update refresh-token related configurations in default config file. So it is like a turn-key solution for developers.

**Comment by jackywxd** [comment]
@daffl David, thanks for creating Feathers! just wondering is there any "contributing guid", "coding guideline", "style guide" for Feathers?

**Comment by daffl** [comment]
As I mentioned before, a PR with the implementation for refresh tokens (or even just some ideas) would be very welcome. I didn't get a chance yet to check out the linked repos and this discussion is getting quite long. Having it all up to date and in one place would make things a lot easier. Some important bullet points:

- Refresh tokens can be issued by [extending the Authentication Service](https://docs.feathersjs.com/api/authentication/service.html#customization)
- Refresh tokens should be stored in localStorage
- Refresh tokens need to be revokable (so there needs to be a more general purpose implementation of the revocation mechanism described in https://docs.feathersjs.com/cookbook/authentication/revoke-jwt.html)
- Because of the additional complexity and setup (like Redis for storing revoked tokens) it can be available as a feature but shouldn't be enabled by default (i.e a standard generated app - can be part of the CLI but before doing anything on that, I'm still really looking for help in starting a [hygen](http://hygen.io/) based generator)
- Contributing info can be found in [the contributors guide](https://github.com/feathersjs/feathers/blob/crow/.github/contributing.md)

**Comment by matiaslopezd** [comment]
If someone is using refreshToken, we create a library that handles in frontend, accessToken and refreshToken like "sessions", too easy to use.

Only need to pass the tokens and the library tries to obtain a new acessToken with refreshToken when will expire. Currently is used in [Videsk](https://videsk.io).

Repository: [Front Auth Handler](https://github.com/videsk/front-auth-handler)

**Comment by J3m5** [comment]
I saw this video about the risks related to JWT andrefresh tokens in SPA, it is really informative.
https://pragmaticwebsecurity.com/talks/xssoauth.html
